### PR TITLE
unifiy constant instruction to support N-D array

### PR DESCRIPTION
**Pull request by CtfGo**
### Describe

1. add `Numel` interface on `Shape` class, which return the number of element
2. modify `ConstantD0` meta_op to `Constant` that supports N-D array also scalar literal now
3. add several template class on `note/type_traits.h` to enable function dispatch on `Constant` according to its data type

### CODE REVIEW

File: paddle/fluid/compiler/piano/note/type_traits.h [hunk 1]
```diff
@@ -224,6 +224,29 @@ class UnboxingIterator : public ForwardIterator<SmartPtrIter> {
   SmartPtrIter iter_;
 };
 
+template <typename T, typename... ALL_T>
```
Comment (Xreki) [inline paddle/fluid/compiler/piano/note/type_traits.h]: 模板里面的变量名是不是要用`AllT`这种命名方式？

**Review by wzzju — DISMISSED** [review]
LGTM

File: paddle/fluid/compiler/piano/note/type_traits.h (hunk 1 above)
Comment (CtfGo) [reply paddle/fluid/compiler/piano/note/type_traits.h]: fix

**Review by Xreki — APPROVED** [review]

**Review by wzzju — APPROVED** [review]
LGTM.

File: paddle/fluid/compiler/piano/symbolization/meta_op_test.cc [hunk 2]
```diff
@@ -32,24 +32,34 @@ TEST(MetaOpTest, TestParameter) {
   EXPECT_EQ(Shape(note::F32, {1, 2}), param_op.Shape());
 }
 
-TEST(MetaOpTest, TestConstantD0) {
-  NoteBuilder builder("test_constant_d0");
-  auto constant_d0_op = ConstantD0<int32_t>(&builder, 110);
+TEST(MetaOpTest, TestConstant) {
+  NoteBuilder builder("test_constant");
+  // add a constant instruction with scalar value
+  auto constant_d0_op = Constant<int32_t>(&builder, 110, Shape(note::S32, {}));
```
Comment (thisjiang) [inline paddle/fluid/compiler/piano/symbolization/meta_op_test.cc:38]: 这样写的结果是不是和`ConstantD0(&builder, 110)`相同？

Comment (CtfGo) [reply paddle/fluid/compiler/piano/symbolization/meta_op_test.cc:38]: 是的

**Pull request merged**